app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/integration/NgVpnNetworkStack.kt

```diff
@@ -0,0 +1,268 @@
+
+    override fun isEnabled(): Boolean {
+        val variant = vpnNetworkStackVariantManager.getVariant()
+        return appTpFeatureConfig.isEnabled(AppTpSetting.VpnNewNetworkingLayer) && variant == name
+    }
+
+    override fun shouldSetActiveNetworkDnsServers(): Boolean {
+        // for native VPN network layer always set active network DNS
+        return shouldSetUnderlyingNetworks()
+    }
+
+    override fun shouldSetUnderlyingNetworks(): Boolean {
+        return appTpFeatureConfig.isEnabled(AppTpSetting.NetworkSwitchHandling)
+    }
+
+    override fun onCreateVpn() {
+        vpnNetwork.addCallback(this)
+
+        if (jniContext != 0L) {
+            vpnNetwork.stop(jniContext)
+            synchronized(jniLock) {
+                vpnNetwork.destroy(jniContext)
+                jniContext = 0
+            }
+
+        }
+        jniContext = vpnNetwork.create()
+    }
+
+    override fun onStartVpn(tunfd: ParcelFileDescriptor) {
+        startNative(tunfd.fd)
+    }
+
+    override fun onStopVpn() {
+        stopNative()
+    }
+
+    override fun onDestroyVpn() {
+        synchronized(jniLock) {
+            vpnNetwork.destroy(jniContext)
+            jniContext = 0
+        }
+        vpnNetwork.addCallback(null)
+    }
+
+    override fun mtu(): Int {
+        return vpnNetwork.mtu()
+    }
+
+    override fun onExit(reason: String) {
+        Timber.w("Native exit reason=$reason")
+        // restart the VPN
+        appCoroutineScope.launch {
+            TrackerBlockingVpnService.restartVpnService(context, forceGc = true)
+        }
+    }
+
+    override fun onError(errorCode: Int, message: String) {
+        Timber.w("onError $errorCode:$message")
+    }
+
+    override fun onDnsResolved(dnsRR: DnsRR) {
+        dnsResourceCache.put(dnsRR.resource, dnsRR.qName)
+        Timber.w("dnsResolved called for $dnsRR")
+    }
+
+    override fun onSniResolved(sniRR: SniRR) {
+        dnsResourceCache.put(sniRR.resource, sniRR.name)
+        Timber.w("sniResolved called for ${sniRR.name} / ${sniRR.resource}")
+    }
+
+    override fun isDomainBlocked(domainRR: DomainRR): Boolean {
+        Timber.w("isDomainBlocked for $domainRR")
+        return !shouldAllowDomain(domainRR.name, domainRR.uid)
+    }
+
+    override fun isAddressBlocked(addressRR: AddressRR): Boolean {
+        val hostname = dnsResourceCache[addressRR.address] ?: return false
+        val domainAllowed = shouldAllowDomain(hostname, addressRR.uid)
+        Timber.w("isAddressBlocked for $addressRR ($hostname) = ${!domainAllowed}")
+        return !domainAllowed
+    }
+
+    private fun shouldAllowDomain(name: String, uid: Int): Boolean {
+        val packageId = getPackageIdForUid(uid)
+
+        if (VpnExclusionList.isDdgApp(packageId)) {
+            Timber.v("shouldAllowDomain: DDG ap is always allowed")
+            return true
+        }
+
+        val type = runBlocking {
+            appTrackerRepository.findTracker(name, packageId)
+        }
```

Comment:
`findTracker` is a `suspend` function, but using `runBlocking` here because anyway all this code runs in its own thread 🤷

Reply:
I don't think the `runBlocking` is needed, everything runs on the same worker thread.

Reply:
oh you're right, `findTrackers` is not a `suspend` fun. Will change